js/compiler/compiler.go

```diff
@@ -119,48 +124,92 @@ func (c *Compiler) Transform(src, filename string) (code string, srcmap []byte,
 		return
 	}
 
-	code, srcmap, err = c.babel.Transform(c.logger, src, filename)
+	code, srcMap, err = c.babel.transformImpl(c.logger, src, filename, c.COpts.SourceMapLoader != nil, inputSrcMap)
 	return
 }
 
+// Options are options to the compiler
+type Options struct {
+	CompatibilityMode lib.CompatibilityMode
+	SourceMapLoader   func(string) ([]byte, error)
+	Strict            bool
+}
+
+// compilationState is helper struct to keep the state of a compilation
+type compilationState struct {
+	// set when we couldn't load external source map so we can try parsing without loading it
+	couldntLoadSourceMap bool
+	// srcMap is the current full sourceMap that has been generated read so far
+	srcMap []byte
+	main   bool
+
+	compiler *Compiler
+}
+
 // Compile the program in the given CompatibilityMode, wrapping it between pre and post code
-func (c *Compiler) Compile(src, filename, pre, post string,
-	strict bool, compatMode lib.CompatibilityMode) (*goja.Program, string, error) {
-	code := pre + src + post
-	ast, err := parser.ParseFile(nil, filename, code, 0, parser.WithDisableSourceMaps)
-	if err != nil {
-		if compatMode == lib.CompatibilityModeExtended {
-			code, _, err = c.Transform(src, filename)
-			if err != nil {
-				return nil, code, err
-			}
-			// the compatibility mode "decreases" here as we shouldn't transform twice
-			return c.Compile(code, filename, pre, post, strict, lib.CompatibilityModeBase)
+func (c *Compiler) Compile(src, filename string, main bool, cOpts Options) (*goja.Program, string, error) {
+	return c.compileImpl(src, filename, main, cOpts, nil)
+}
+
+// here we take the srcMap as pointer so we can change in case that we have an original srcMap that will be loaded, but
+// the code still needs to go through babel. In this way we can give the correct inputSourceMap to babel which hopefully
+// will generate a sourcemap which goes from the final source file lines to the original ones that were transpiled
+// outside of k6
+// TODO have some kind of "compilationContext" to save this instead and modify and move that
+// This won't be needed if we weren't going through babel so removing babel would kind of remove the need for that
+func (c *compilationState) sourceMapLoader(path string) ([]byte, error) {
+	if path == sourceMapURLFromBabel {
+		if !c.main {
+			return increaseMappingsByOne(c.srcMap)
 		}
-		return nil, code, err
+		return c.srcMap, nil
+	}
+	var err error
+	c.srcMap, err = c.compiler.COpts.SourceMapLoader(path)
+	if err != nil {
+		c.couldntLoadSourceMap = true
+		return nil, err
+	}
+	if !c.main {
+		return increaseMappingsByOne(c.srcMap)
+	}
+	return c.srcMap, err
+}
+
+func (c *Compiler) compileImpl(
+	src, filename string, main bool, cOpts Options, srcMap []byte,
+) (*goja.Program, string, error) {
+	code := src
+	state := compilationState{srcMap: srcMap, compiler: c, main: main}
+	if !main {
+		code = "(function(module, exports){\n" + code + "\n})\n"
```

Comment:
doesn't this affect the source map? :confused:

Reply:
this is what https://github.com/grafana/k6/pull/2082/files#r782919998 fixes ;)

Reply:
that isn't obvious from the code or the comments :man_shrugging:

Reply:
is  44c2849  sufficient?